### ISSUE TITLE
Proposal: Function as `supplement`

### ISSUE THREAD

**Issue by MDLC01** [typst/typst#2990]
### Description

Many elements have a `supplement` parameter. While it is useful, the fact that it only lets you customize what is displayed before a number is limiting.

We should allow `supplement`s to be functions that accept the formatted number (as content), and return some content that contains the number with its supplement. For example, `x => [Before #x after.]`.

```typ
#figure(
  supplement: n => [The #n#super[th] figure],
  caption: [A circle.],
  circle(),
)
```
![A circle with the legend "The 8th figure: A circle."](https://github.com/typst/typst/assets/57839069/3a948cd7-d9ca-4382-9c2a-7aebe3ed3e87)

### Use Case

Easier customization. This also makes it possible to remove the space between the supplement and the number.

### Current alternative

This can already be done by adding the content you want to have after the number within the numbering format, but this is far from optimal.

```typ
#figure(
  supplement: [],
  numbering: n => [The #n#super[th] figure],
  caption: [A circle.],
  circle(),
)
```

**Comment by Dherse** [comment]
Be careful with just `th` as the super otherwise you'll get `21th` or `103th` which just looks weird!

**Comment by MDLC01** [comment]
> Be careful with just `th` as the super otherwise you'll get `21th` or `103th` which just looks weird!

Yes of course. This was just the first example I could think of.

**Comment by MDLC01** [comment]
I just realized that this proposal can't be implemented this way, because supplements can already be functions that accept the element's body.

I think the underlying feature request is still interesting, though.

**Comment by wrzian** [comment]
What if we allowed an extension to the supplement syntax: if you type `@label[custom supp]` then we allow you to use `<label>` as though it were a variable in the inner markup block? So you could write: `@thm:A[Theorem <thm:A> by Author]`.

If you don't include `<label>` then we can make the supplement default to prefix of the reference, or add a set rule that changes it to postfix.

**Comment by laurmaedje** [comment]
Making the label suddenly evaluate to content seems strange to me, to be honest. It's not really the principle of least surprise.

**Comment by wrzian** [comment]
Fair. I chose that because placing the label again doesn't have any current meaning. Some alternatives if this is desirable might be repeating `@label` or maybe using just a raw `@` or `#@`.
I do think this style has potential to solve the context-dependent supplement problem.

**Comment by laurmaedje** [comment]
Is the need for this single-ref context-dependance really that common that it deserves custom syntax? I'd imagine that you'd typically either rather want to configure this in a set rule (as a function), not ad-hoc for a single reference.

**Comment by YDX-2147483647** [comment]
> I just realized that this proposal can't be implemented this way, because supplements can already be functions that accept the element's body. (https://github.com/typst/typst/issues/2990#issuecomment-1859217295)

Could the issue title be modified for future reference? For example:

> Change the return value of `supplement` functions from `Figure` to `Figure 1`.

- **Current:** (a `figure` element) ↦ `[Figure]`.
- **Possible changes:**
   - `(1,)` ↦ `[Figure 1]`
   - (a `figure` element,) ↦ `[Figure 1]`
   - (a `figure` element, `1`) ↦ `[Figure 1]`
   - (a `figure` element, supplement: `[Fig.]`) ↦ `[Fig. 1]`, where `[Fig.]` is given by writing `@key[Fig.]`.
   - …


## Update on 2026-04-17

The proposed `{figure,heading,…}.supplement` function API has an advantage over [`show ref: it => {…}` customization](https://typst.app/docs/reference/model/ref/#customization):

In a `show ref` function, we have to recreate a `link`.
If `link` and `ref` have different styles (e.g., different text color), then the style for the created `link` will override the style for `ref`, which is not desired.

Example:

```typst
#show link: set text(blue.darken(20%))
#show ref: set text(green.darken(30%))

#figure(rect[Image], caption: [Caption]) <a>

= Default
/ `ref`: @a
/ `link`: https://example.com

= Customized
#show ref: it => {
  let el = it.element
  if el != none and el.func() == figure {
    link(el.location(), {
      set text(green.darken(30%))
      // 👆 Without this, the `ref` will become blue.
      [第~]
      numbering(el.numbering, ..el.counter.at(el.location()))
      [~图]
    })
  } else {
    it
  }
}
/ `ref`: @a
/ `link`: https://example.com
```

<img width="350" alt="Image" src="https://github.com/user-attachments/assets/81cee006-0a7d-4660-a04e-c3b9fb39c4d8" />
